daemon/cmd/devices.go

```diff
@@ -0,0 +1,257 @@
+			for dev := range dm.devices {
+				link, err := netlink.LinkByName(dev)
+				if err != nil {
+					return fmt.Errorf("Cannot find device '%s': %w", dev, err)
+				}
+
+				// Check if any of the addresses assigned to this device is the k8s node IP and if so
+				// use it for direct routing and IPv6 multicast.
+				if addrs, err := netlink.AddrList(link, netlink.FAMILY_ALL); err == nil {
+					for _, a := range addrs {
+						if a.IP.Equal(nodeIP) {
+							if detectDirectRoutingDev {
+								option.Config.DirectRoutingDevice = dev
+								log.Infof("Detected %s=%s", option.DirectRoutingDevice, option.Config.DirectRoutingDevice)
+							}
+							if detectIPv6MCastDev && link.Attrs().Flags&net.FlagMulticast != 0 {
+								option.Config.IPv6MCastDevice = dev
+								log.Infof("Detected %s=%s", option.IPv6MCastDevice, option.Config.IPv6MCastDevice)
+							}
+							break
+						}
+					}
+				} else {
+					log.WithError(err).Warnf("Cannot retrieve device '%s' IP addresses, skipping it", dev)
+				}
+			}
+
+			if detectDirectRoutingDev && option.Config.DirectRoutingDevice == "" {
+				return fmt.Errorf("Unable to determine BPF NodePort direct routing device. "+
+					"Use --%s to specify it", option.DirectRoutingDevice)
+			}
+			if detectIPv6MCastDev && option.Config.IPv6MCastDevice == "" {
+				return fmt.Errorf("Unable to determine Multicast device. Use --%s to specify them",
+					option.IPv6MCastDevice)
+			}
+		}
+	}
+
+	option.Config.Devices = dm.getDevices()
+	log.WithField(logfields.Devices, option.Config.Devices).Info("Detected devices")
+
+	return nil
+}
+
+func (dm *DeviceManager) GetDevices() []string {
+	dm.mu.Lock()
+	defer dm.mu.Unlock()
+	return dm.getDevices()
+}
+
+func (dm *DeviceManager) getDevices() []string {
+	devs := make([]string, 0, len(dm.devices))
+	for dev := range dm.devices {
+		devs = append(devs, dev)
+	}
+	sort.Strings(devs)
+	return devs
+}
+
+// isViableDevice returns true if the given link is usable and Cilium should attach
+// programs to it.
+func (dm *DeviceManager) isViableDevice(link netlink.Link) bool {
+	name := link.Attrs().Name
+
+	// Do not consider any of the excluded devices.
+	for _, p := range excludedDevicePrefixes {
+		if strings.HasPrefix(name, p) {
+			log.WithField(logfields.Device, name).Debugf("Skipping device as it has excluded prefix '%s'", p)
+			return false
+		}
+	}
+
+	// Skip slave devices.
+	if link.Attrs().RawFlags&unix.IFF_SLAVE != 0 {
+		log.WithField(logfields.Device, name).Debugf("Skipping slave device")
+		return false
+	}
+
+	// Ignore L3 devices if we cannot support them.
+	if !dm.l3DevOK && !mac.LinkHasMacAddr(link) {
+		log.WithField(logfields.Device, name).
+			Warn("Ignoring L3 device; >= 5.8 kernel is required.")
+		return false
+	}
+
+	return true
+}
+
+// updateDevicesFromRoutes processes a batch of routes and updates the set of
+// devices. Returns true if devices changed.
+func (dm *DeviceManager) updateDevicesFromRoutes(routes []netlink.Route) bool {
```

Comment:
A follow-up PR will reuse this logic to detect devices at runtime and reload the datapath.